### PR TITLE
Feature: Allow re-use of run_test() in install_time_test_callbacks

### PR DESCRIPTION
**Pull request by tldahlgren**
@adamjstewart @bernhardkaindl 

This PR supports re-use of `PackageBase`'s `run_test` method within `install_time_test_callbacks`, addressing one of two issues raised in #26575 discussions.  (The import-tests change discussed in the PR are intended for a separate PR unless someone objects.)

The changes:
- Ensures key package properties are initialized when processing `install_time_test_callbacks`;
- Directs debug-level output from the associated methods -- especially when using `run_test` -- to a separate install test log;
- Ensures the build log contains an entry about running the tests and, if failure(s) occur, the error output; and
- Copies the install test log to the install directory.

### Updated output using `py-ats` for callback testing
```
$  spack install --verbose --test root py-ats
...
==> Installing py-ats-7.0.100-beqy7pd3p3eo674v64mtwonwd2hq4jrr
==> No binary for py-ats-7.0.100-beqy7pd3p3eo674v64mtwonwd2hq4jrr found: installing from source
...
Successfully installed ats-7.0.100
Removed build tracker: '/tmp/dahlgren/pip-req-tracker-40r2u_r4'
==> [2022-03-10-14:54:05.090787] Running install-time tests
==> Testing package py-ats-7.0.100-beqy7pd
==> [2022-03-10-14:54:05.107120] RUN-TESTS: install-time tests [test]
==> [2022-03-10-14:54:05.124956] Detected the following modules: ['ats', 'ats.atsMachines', 'ats.bin', 'ats.util']
==> [2022-03-10-14:54:05.125819] checking import of ats
==> [2022-03-10-14:54:05.126419] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats'
PASSED
==> [2022-03-10-14:54:05.286421] checking import of ats.atsMachines
==> [2022-03-10-14:54:05.287073] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats.atsMachines'
PASSED
==> [2022-03-10-14:54:05.442490] checking import of ats.bin
==> [2022-03-10-14:54:05.443249] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats.bin'
PASSED
==> [2022-03-10-14:54:05.609549] checking import of ats.util
==> [2022-03-10-14:54:05.610268] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats.util'
PASSED
==> py-ats: Successfully installed py-ats-7.0.100-beqy7pd3p3eo674v64mtwonwd2hq4jrr
  Fetch: 0.04s.  Build: 7.99s.  Total: 8.02s.
[+] /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/py-ats-7.0.100-beqy7pd3p3eo674v64mtwonwd2hq4jrr
700.772u 112.733s 10:14.65 132.3%	0+0k 989312+2952152io 546pf+0w
```

And running stand-alone tests:

```
$ spack -v test run py-ats
==> Spack test 3umzj3oc7p36rpq72bxpgfwmnkmh57kw
==> Testing package py-ats-7.0.100-beqy7pd
==> [2022-03-10-14:56:36.837772] Detected the following modules: ['ats', 'ats.atsMachines', 'ats.bin', 'ats.util']
==> [2022-03-10-14:56:36.839711] checking import of ats
==> [2022-03-10-14:56:36.840235] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats'
PASSED
==> [2022-03-10-14:56:37.023600] checking import of ats.atsMachines
==> [2022-03-10-14:56:37.024224] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats.atsMachines'
PASSED
==> [2022-03-10-14:56:37.142730] checking import of ats.bin
==> [2022-03-10-14:56:37.143321] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats.bin'
PASSED
==> [2022-03-10-14:56:37.260456] checking import of ats.util
==> [2022-03-10-14:56:37.260959] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/python-3.9.10-i6t33rdskzbrynmczxjovk4xmdzwtjzn/bin/python3.9' '-c' 'import ats.util'
PASSED
============================= 1 passed of 1 specs ==============================
```

### Old Examples: Failed builds
Example output at the tail of a failing build log:
```
==> [2021-10-07-19:38:58.010952] Running install-time tests
==> [2021-10-07-19:39:33.028145] 
1 error found in test log:
     112      File "<string>", line 1, in <module>
     113      File "/tmp/dahlgren/spack-stage/spack-stage-py-numpy-1.21.2-eegaz
            7jqduhpt3xpp67l4p5boj6ue4h6/spack-src/numpy/__init__.py", line 132,
             in <module>
     114        raise ImportError(msg) from e
     115    ImportError: Error importing numpy: you should not try to import nu
            mpy from
     116            its source directory; please exit the numpy source tree, an
            d relaunch
     117            your python interpreter from there.
  >> 118    FAILED: Command exited with status 1:
     119        '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/g
            cc-8.3.1/python-3.9.6-kwzsf3lkqaq5zhmblkwtqphulymteuos/bin/python3.
            9' '-c' 'import numpy; numpy.test(verbose=2)'
     120    Traceback (most recent call last):
     121      File "/tmp/dahlgren/spack-stage/spack-stage-py-numpy-1.21.2-eegaz
            7jqduhpt3xpp67l4p5boj6ue4h6/spack-src/numpy/__init__.py", line 127,
             in <module>
     122        from numpy.__config__ import show as show_config
     123    ModuleNotFoundError: No module named 'numpy.__config__'
     124    
```

And output from the corresponding install test log:
```
==> [2021-10-07-19:38:58.027254] RUN-TESTS: install-time tests [test]
==> [2021-10-07-19:38:58.126790] Detected the following modules: ['numpy', 'numpy.compat', 'numpy.compat.tests', 'numpy.core', 'numpy.core.tests', 'numpy.distutils', 'numpy.distutils.command', 'numpy.distutils.fcompiler', 'numpy.distutils.tests', 'numpy.doc', 'numpy.f2py', 'numpy.f2py.tests', 'numpy.fft', 'numpy.fft.tests', 'numpy.lib', 'numpy.lib.tests', 'numpy.linalg', 'numpy.linalg.tests', 'numpy.ma', 'numpy.ma.tests', 'numpy.matrixlib', 'numpy.matrixlib.tests', 'numpy.polynomial', 'numpy.polynomial.tests', 'numpy.random', 'numpy.random.tests', 'numpy.random.tests.data', 'numpy.testing', 'numpy.testing._private', 'numpy.testing.tests', 'numpy.typing', 'numpy.typing.tests', 'numpy.tests']
==> [2021-10-07-19:38:58.127364] checking import of numpy
==> [2021-10-07-19:38:58.127751] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/python-3.9.6-kwzsf3lkqaq5zhmblkwtqphulymteuos/bin/python3.9' '-c' 'import numpy'
PASSED
==> [2021-10-07-19:39:00.270122] checking import of numpy.compat
==> [2021-10-07-19:39:00.270639] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/python-3.9.6-kwzsf3lkqaq5zhmblkwtqphulymteuos/bin/python3.9' '-c' 'import numpy.compat'
PASSED
[..snip..]
==> [2021-10-07-19:39:32.611902] checking import of numpy.tests
==> [2021-10-07-19:39:32.612377] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/python-3.9.6-kwzsf3lkqaq5zhmblkwtqphulymteuos/bin/python3.9' '-c' 'import numpy.tests'
PASSED
==> [2021-10-07-19:39:32.920791] test: running numpy.test
==> [2021-10-07-19:39:32.921231] '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/python-3.9.6-kwzsf3lkqaq5zhmblkwtqphulymteuos/bin/python3.9' '-c' 'import numpy; numpy.test(verbose=2)'
Traceback (most recent call last):
  File "/tmp/dahlgren/spack-stage/spack-stage-py-numpy-1.21.2-eegaz7jqduhpt3xpp67l4p5boj6ue4h6/spack-src/numpy/__init__.py", line 127, in <module>
    from numpy.__config__ import show as show_config
ModuleNotFoundError: No module named 'numpy.__config__'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/tmp/dahlgren/spack-stage/spack-stage-py-numpy-1.21.2-eegaz7jqduhpt3xpp67l4p5boj6ue4h6/spack-src/numpy/__init__.py", line 132, in <module>
    raise ImportError(msg) from e
ImportError: Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there.
FAILED: Command exited with status 1:
    '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/python-3.9.6-kwzsf3lkqaq5zhmblkwtqphulymteuos/bin/python3.9' '-c' 'import numpy; numpy.test(verbose=2)'
Traceback (most recent call last):
  File "/tmp/dahlgren/spack-stage/spack-stage-py-numpy-1.21.2-eegaz7jqduhpt3xpp67l4p5boj6ue4h6/spack-src/numpy/__init__.py", line 127, in <module>
    from numpy.__config__ import show as show_config
ModuleNotFoundError: No module named 'numpy.__config__'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/tmp/dahlgren/spack-stage/spack-stage-py-numpy-1.21.2-eegaz7jqduhpt3xpp67l4p5boj6ue4h6/spack-src/numpy/__init__.py", line 132, in <module>
    raise ImportError(msg) from e
ImportError: Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there.

  File "/usr/WS1/dahlgren/spack/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/main.py", line 814, in main
    return _invoke_command(command, parser, args, unknown)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/main.py", line 527, in _invoke_command
    return_val = command(parser, args)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/cmd/install.py", line 473, in install
    install_specs(args, kwargs, zip(abstract_specs, specs))
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/cmd/install.py", line 274, in install_specs
    builder.install()
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/installer.py", line 1616, in install
    self._install_task(task)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/installer.py", line 1169, in _install_task
    pkg, build_process, install_args)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/build_environment.py", line 1122, in start_build_process
    p.start()
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/process.py", line 112, in start
    self._popen = self._Popen(self)
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/context.py", line 223, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/popen_fork.py", line 74, in _launch
    code = process_obj._bootstrap()
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/tce/packages/python/python-3.7.2/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/build_environment.py", line 1021, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/installer.py", line 1918, in build_process
    return installer.run()
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/installer.py", line 1784, in run
    self._real_install()
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/installer.py", line 1882, in _real_install
    phase(pkg.spec, pkg.prefix)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/package.py", line 123, in phase_wrapper
    callback(instance)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/package.py", line 414, in _wrapper
    func(instance, *args, **kwargs)
  File "/usr/WS1/dahlgren/spack/lib/spack/spack/package.py", line 2602, in _run_default_install_time_test_callbacks
    fn()
  File "/usr/WS1/dahlgren/spack/var/spack/repos/builtin/packages/py-numpy/package.py", line 391, in test
    purpose="test: running numpy.test"
```

### CODE REVIEW

**Comment by bernhardkaindl** [comment]
@ looks not too bad. Two observations though:

If stdout of spack is a pipe or a file, each failure is sent twice:
```
spack install --test=root py-hypothesis 2>&1 | grep 'No module'
ModuleNotFoundError: No module named 'numpy'
     17    ModuleNotFoundError: No module named 'numpy'
ModuleNotFoundError: No module named 'django'
     82    ModuleNotFoundError: No module named 'django'
```

But, when I show the output on the tty directly, only the 1st of the two failures is shown twice.

With -v/--verbose, each error is shown 3 times:
```
spack install --test=root -v py-hypothesis 2>&1 | tee log
     17    ModuleNotFoundError: No module named 'numpy'
     82    ModuleNotFoundError: No module named 'django'
ModuleNotFoundError: No module named 'numpy'
     17    ModuleNotFoundError: No module named 'numpy'
ModuleNotFoundError: No module named 'django'
     82    ModuleNotFoundError: No module named 'django'
```
My solution in PR #26575 only showed each failure once when output was a tty and --verbose wasn't used.

It didn't attempt to duplicate the test errors to the build log as well, but that does not appear to be required for me, and it was much simpler.

It would be good to be able enable the import tests even by default when building python packages:
Failing imports are quite common (I've found 8 lately which I haven't submitted yet...)

**Comment by tldahlgren** [comment]
> @ looks not too bad. Two observations though:
> 
> If stdout of spack is a pipe or a file, each failure is sent twice:
> 
> ```
> spack install --test=root py-hypothesis 2>&1 | grep 'No module'
> ModuleNotFoundError: No module named 'numpy'
>      17    ModuleNotFoundError: No module named 'numpy'
> ModuleNotFoundError: No module named 'django'
>      82    ModuleNotFoundError: No module named 'django'
> ```
> 
> But, when I show the output on the tty directly, only the 1st of the two failures is shown twice.
> 
> With -v/--verbose, each error is shown 3 times:
> 
> ```
> spack install --test=root -v py-hypothesis 2>&1 | tee log
>      17    ModuleNotFoundError: No module named 'numpy'
>      82    ModuleNotFoundError: No module named 'django'
> ModuleNotFoundError: No module named 'numpy'
>      17    ModuleNotFoundError: No module named 'numpy'
> ModuleNotFoundError: No module named 'django'
>      82    ModuleNotFoundError: No module named 'django'
> ```
> 
> My solution in PR #26575 only showed each failure once when output was a tty and --verbose wasn't used.
> 
> It didn't attempt to duplicate the test errors to the build log as well, but that does not appear to be required for me, and it was much simpler.

I hadn't considered duplicate error messages.  I wanted to make sure we were capturing the information in a log and, more specifically, the test log file.  IIRC, during testing, I found the test log file we are initializing ends up empty (hence the `with`) while the errors were going to the build log.

I debated whether that IS the preferred approach.  But, if it is, the need for initializing the test log to enable re-use `run_test()` is a hack as the method was not designed to be used in the install scope.

> 
> It would be good to be able enable the import tests even by default when building python packages: Failing imports are quite common (I've found 8 lately which I haven't submitted yet...)

I did not address the imports issue in this PR.  I agree it's a good sanity check.  We do support those at build time though, at this point, the only such checks I recall us supporting are for the presence of files and directories.

**Comment by tldahlgren** [comment]
@adamjstewart @bernhardkaindl @tgamblin @becker33 

Perhaps re-using `run_test` is NOT what is needed for import tests.

I'd like to propose changing those checks from being implemented in `def test` -- so `test` and `run_test` continue to be reserved for stand-alone tests -- to being a new `sanity check` that is executed at install time that could optionally be run at stand-alone time.

**Comment by bernhardkaindl** [comment]
> @adamjstewart @bernhardkaindl @tgamblin @becker33
> 
> Perhaps re-using `run_test` is NOT what is needed for import tests.
> 
> I'd like to propose changing those checks from being implemented in `def test` -- so `test` and `run_test` continue to be reserved for stand-alone tests -- to being a new `sanity check` that is executed at install time that could optionally be run at stand-alone time.

For a long term solution, it would make sense  to automatically have a stand-alone import test which can run @ install time as well (even enabled by default).

For the short term, it would look better for spack users, if the error message from a failed import test would be more user-friendly than this - This what we create as output at the moment:

```py
==> Installing py-slurm-pipeline-3.0.2-svrkymo2mwrezwtxch4idvk7s5mc6lvj
==> No binary for py-slurm-pipeline-3.0.2-svrkymo2mwrezwtxch4idvk7s5mc6lvj found: installing from source
==> Using cached archive: /home/kaindlb/home/spack/.dl/_source-cache/archive/28/28e07eb93e846b395a16e6778fd3fc8344a82d115a6a8420276ec68f67f7131c.tar.gz
==> No patches needed for py-slurm-pipeline
==> py-slurm-pipeline: Executing phase: 'build'
==> py-slurm-pipeline: Executing phase: 'install'
==> Error: AttributeError: 'PySlurmPipeline' object has no attribute 'test_log_file'

/home/kaindlb/home/spack/lib/spack/spack/package.py:1874, in run_test:
       1871                if spack.config.get('config:fail_fast', False):
       1872                    raise TestFailure([(exc, m)])
       1873                else:
  >>   1874                    self.test_failures.append((exc, m))
       1875                return False
```
There isn't even a word that the cause is a missing dependency found by an import check, It shows a internal bug in spack.

@adamjstewart @bernhardkaindl @tgamblin @becker33

My PR #26575 fixes this for the short-term nicely and clearly until the long-term solution is found.

**Comment by tldahlgren** [comment]
@tgamblin We discussed this in the Monday meeting.  Did you want changes to this PR before it can be merged?

**Comment by alalazo** [comment]
Moved to the next point release as it currently needs a rebase + review.

File: var/spack/repos/builtin.mock/packages/test-missing-test/package.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+from spack.package import run_after
+
+
+class TestMissingTest(Package):
```
Comment (becker33) [inline var/spack/repos/builtin.mock/packages/test-missing-test/package.py]: I don't understand what the missinig test method has to do with the test using this package... was this repurposed from some other work?

File: lib/spack/spack/package.py [hunk 2]
```diff
@@ -2607,15 +2620,39 @@ def _run_default_install_time_test_callbacks(self):
...
-                msg = 'RUN-TESTS: method not implemented [{0}]'
-                tty.warn(msg.format(name))
-            else:
-                tty.msg('RUN-TESTS: install-time tests [{0}]'.format(name))
-                fn()
+        # Ensure capture of output for tests that use run_test
+        self.test_failures = []
+        self.test_log_file = fsys.join_path(
+            self.stage.path, _spack_install_test_log)
+        fsys.touch(self.test_log_file)  # Otherwise log_parse complains
```
Comment (becker33) [inline lib/spack/spack/package.py]: This, along with the `log_output` call and the `if self.test_failures` block, feel like they should be factored into a context manager. Could probably be reused in the normal `spack test run` codepath as well.

**Review by becker33 — CHANGES_REQUESTED** [review]
Is the idea that this should work for install-time test callbacks, but not build-time test callbacks because the build-time callbacks can't be run against the installed object?

File: var/spack/repos/builtin.mock/packages/test-missing-test/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin.mock/packages/test-missing-test/package.py]: It's been a while but I am assuming it was for test coverage.

File: lib/spack/spack/package.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/package.py]: I'll look into that.  Thanks for the feedback.

**Comment by tldahlgren** [comment]
> Is the idea that this should work for install-time test callbacks, but not build-time test callbacks because the build-time callbacks can't be run against the installed object?

The problem, as I recall, is that there are packages that want or have included `test` in `install_time_test_callbacks` and or tried to use `run_test()` in their post-install phase tests but get failures from it that are't clear.  I had not considered precluding use after the `build` phase but that would probably make sense in most cases.

I see this PR as a prerequisite to Adam's recent comments/request to automatically run `test` after the `install` phase.

File: lib/spack/spack/package.py [hunk 3]
```diff
@@ -1816,13 +1829,10 @@ def do_test(self, dirty=False):
...
-        self.test_log_file = self.test_suite.log_file_for_spec(self.spec)
-        self.tested_file = self.test_suite.tested_file_for_spec(self.spec)
-        fsys.touch(self.test_log_file)  # Otherwise log_parse complains
-
-        kwargs = {'dirty': dirty, 'fake': False, 'context': 'test'}
-        spack.build_environment.start_build_process(self, test_process, kwargs)
+        with self._setup_test():
+            kwargs = {'dirty': dirty, 'fake': False, 'context': 'test'}
+            spack.build_environment.start_build_process(
+                self, test_process, kwargs)
```
Comment (becker33) [inline lib/spack/spack/package.py]: I think if you push the context manager inside the test_process method, then you can include error handling in the context manager instead of having to handle it separately at the end of `install_time_callbacks`

Comment (tldahlgren) [reply lib/spack/spack/package.py]: The flow in `test_process` differs a bit from `_run_default_install_time_test_callbacks` in terms of options and flagging test state.

Are you also suggesting embedding use of the `log_output` context manager in `_setup_test()`?

File: lib/spack/spack/package.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/package.py]: Done

File: lib/spack/spack/package.py (hunk 3 above)
Comment (tldahlgren) [reply lib/spack/spack/package.py]: Nvm.  I pushed it too anyway.  Provides consistent log output (see updated description).

**Comment by tldahlgren** [comment]
@becker33 Rebased to restart the checks after some odd bootstrap and pipeline failures.  The checks are passing now.

File: lib/spack/spack/package.py [hunk 4]
```diff
@@ -2607,27 +2647,32 @@ def _run_default_install_time_test_callbacks(self):
...
+                except AttributeError:
+                    msg = 'RUN-TESTS: method not implemented [{0}]'
+                    tty.warn(msg.format(name))
+                else:
+                    tty.msg('RUN-TESTS: install-time tests [{0}]'
+                            .format(name))
+                    try:
+                        fn()
+                    except BaseException as exc:
+                        self.test_failures.append((exc, str(exc)))
```
Comment (becker33) [inline lib/spack/spack/package.py]: Shouldn't the underlying method (`fn`) call `pkg.run_test` which handles this exception catching? Am I missing something?

File: lib/spack/spack/package.py [hunk 5]
```diff
@@ -2607,27 +2647,32 @@ def _run_default_install_time_test_callbacks(self):
...
 
-def test_process(pkg, kwargs):
-    with tty.log.log_output(pkg.test_log_file) as logger:
-        with logger.force_echo():
-            tty.msg('Testing package {0}'
-                    .format(pkg.test_suite.test_pkg_id(pkg.spec)))
+        # Make sure the build log contains the failure(s)
+        for _, msg in self.test_failures:
+            tty.msg(msg)
```
Comment (becker33) [inline lib/spack/spack/package.py]: Why is this needed here but not in `test_process`? Having trouble parsing this.

File: lib/spack/spack/package.py (hunk 4 above)
Comment (tldahlgren) [reply lib/spack/spack/package.py]: This is the loop for post-install build-time tests, i.e., running the methods explicitly listed in `install_time_test_callbacks`. So they are running package methods.  The `run_test` method takes the name or path of an executable.

Some packages are already including the `test` method when they override the list:
- python
- sip
- py-anuga
- py-datalad

File: lib/spack/spack/package.py (hunk 5 above)
Comment (tldahlgren) [reply lib/spack/spack/package.py]: The callbacks are performed after the install-phase and, as I mention in the comment at line 2669, this ensures the test messages are reflected in the build log in case that's all someone thinks to look at since it is part of the installation process.

We don't need it in `test_process`, because that method is associated with `do_test`, tied to the stand-alone test process, and uses different logging.  (Note `do_test` is the method used by reporters to capture stand-alone test output.)

**Comment by tldahlgren** [comment]
@becker33 @tgamblin @alalazo Are there any other concerns/issues with this PR I need to address?

**Comment by tldahlgren** [comment]
@becker33 @tgamblin Are there any more changes needed for this PR?

**Comment by tldahlgren** [comment]
@becker33 @tgamblin Are there any more changes needed for this PR? 

While patch coverage is above target, it appears project coverage has surprisingly taken a hit.

**Comment by adamjstewart** [comment]
Usually rebasing on develop fixes changes to project coverage.

**Comment by becker33** [comment]
@tldahlgren tests are still failing

**Comment by tldahlgren** [comment]
> @tldahlgren tests are still failing

I had to manually fix a rebase to develop but did not address associated unit tests.

**Comment by tldahlgren** [comment]
@becker33 Looks like the post-rebase fixes work and patch coverage is at 92.86%.

**Comment by tldahlgren** [comment]
@becker33 @tgamblin Are there any changes (e.g., approach) for this PR?

File: lib/spack/spack/package.py [hunk 6]
```diff
@@ -2601,6 +2636,9 @@ def _run_default_build_time_test_callbacks(self):
         if self.build_time_test_callbacks is None:
             return
 
+        # Log running install-time tests to the build log
+        tty.msg('Running install-time tests')
+
```
Comment (becker33) [inline lib/spack/spack/package.py]: Message is in the wrong method

**Review by becker33 — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin.mock/packages/test-missing-test/package.py (hunk 1 above)
Comment (becker33) [reply var/spack/repos/builtin.mock/packages/test-missing-test/package.py]: Maybe rename to `TestInstallCallbackTest`

File: lib/spack/spack/package.py (hunk 6 above)
Comment (tldahlgren) [reply lib/spack/spack/package.py]: Actually, that was a copy-and-paste issue.  I meant but forgot to change it.

**Comment by tldahlgren** [comment]
@becker33 @tgamblin Any chance of getting a review on this before it requires yet another (possibly painful 😉 ) rebase and clean up?

File: lib/spack/spack/package.py [hunk 7]
```diff
@@ -2694,45 +2729,56 @@ def rpath_args(self):
         """
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
+    def _run_test_callbacks(self, method_names, build_time=False):
+        """Tries to call all of the listed methods, returning immediately
+           if the list is None."""
+        if method_names is None:
+            return
+
+        callback_desc = 'build' if build_time else 'install'
```
Comment (becker33) [inline lib/spack/spack/package.py]:
```suggestion
    def _run_test_callbacks(self, method_names, callback_type='install'):
        """Tries to call all of the listed methods, returning immediately
           if the list is None."""
        if method_names is None:
            return

```
I think something like this would make things more legible.

File: lib/spack/spack/package.py [hunk 8]
```diff
@@ -2694,45 +2729,56 @@ def rpath_args(self):
...
+        fail_fast = spack.config.get('config:fail_fast', False)
+        externals, verbose = False, False
+
+        with self._setup_test(verbose, externals) as logger:
```
Comment (becker33) [inline lib/spack/spack/package.py]:
```suggestion

        with self._setup_test(verbose=False, externals=False) as logger:
```

I think this is easier to read as well, let me know if you disagree

File: lib/spack/spack/package.py [hunk 9]
```diff
@@ -2839,23 +2878,24 @@ def test_process(pkg, kwargs):
 
                     # Run the tests
                     ran_actual_test_function = True
-                    test_fn(pkg)
+                    if verbose:
+                        with logger.force_echo():
+                            test_fn(pkg)
+                    else:
+                        test_fn(pkg)
```
Comment (becker33) [inline lib/spack/spack/package.py]:
```suggestion
                    context = logger.force_echo if verbose else nullcontext
                    with context:
                        test_fn(pkg)
```

Do you think this would be easier or harder to maintain? Having two main entrypoints to the method gives me the willies, but I'm not sure this is better.

**Review by becker33 — CHANGES_REQUESTED** [review]
This is looking very good, just some style questions/suggestions, feel free to push back.

**Review by becker33 — APPROVED** [review]
